### ISSUE TITLE
Unable to view AngularJS 2.0 Beta 1 in Internet Explorer: The use of a keyword for an identifier is invalid

### ISSUE THREAD

**Issue by kneuse** [angular/angular#6501]
I am unable to render AngularJS 2.0 Beta 1 in Internet Explore. 

I have stripped down a page to as small as possible and can reproduce the issue where all I have as an empty app component that references an html page. 

Downgrading to Angular JS 2.0 Beta 0 resolves the issue. 

Below is the stack trace of the error:

HTML1300: Navigation occurred.
File: localhost:8000
EXCEPTION: SyntaxError: The use of a keyword for an identifier is invalid
   EXCEPTION: SyntaxError: The use of a keyword for an identifier is invalid
   STACKTRACE:
   SyntaxError: The use of a keyword for an identifier is invalid
   at evalExpression (http://localhost:8000/lib/js/angular2.dev.js:452:5)
   at TemplateCompiler.prototype._createViewFactoryRuntime (http://localhost:8000/lib/js/angular2.dev.js:23969:9)
   at Anonymous function (http://localhost:8000/lib/js/angular2.dev.js:23924:13)
   at run (http://localhost:8000/lib/js/angular2-polyfills.js:138:7)
   at Anonymous function (http://localhost:8000/lib/js/angular2.dev.js:13247:15)
   at zoneBoundFn (http://localhost:8000/lib/js/angular2-polyfills.js:111:7)
   at lib$es6$promise$$internal$$tryCatch (http://localhost:8000/lib/js/angular2-polyfills.js:1511:9)
   at lib$es6$promise$$internal$$invokeCallback (http://localhost:8000/lib/js/angular2-polyfills.js:1523:9)
   at Anonymous function (http://localhost:8000/lib/js/angular2-polyfills.js:2071:13)
   at Anonymous function (http://localhost:8000/lib/js/angular2-polyfills.js:243:5)

-----async gap-----
Error
   at _getStacktraceWithCaughtError (http://localhost:8000

SyntaxError: The use of a keyword for an identifier is invalid
   {
      [functions]: ,
      __proto__: { },
      description: "The use of a keyword for an identifier is invalid",
      message: "The use of a keyword for an identifier is invalid",
      name: "SyntaxError",
      number: -2146827240,
      stack: "SyntaxError: The use of a keyword for an identifier is invalid
   at evalExpression (http://localhost:8000/lib/js/angular2.dev.js:452:5)
   at TemplateCompiler.prototype._createViewFactoryRuntime (http://localhost:8000/lib/js/angular2.dev.js:23969:9)
   at Anonymous function (http://localhost:8000/lib/js/angular2.dev.js:23924:13)
   at run (http://localhost:8000/lib/js/angular2-polyfills.js:138:7)
   at Anonymous function (http://localhost:8000/lib/js/angular2.dev.js:13247:15)
   at zoneBoundFn (http://localhost:8000/lib/js/angular2-polyfills.js:111:7)
   at lib$es6$promise$$internal$$tryCatch (http://localhost:8000/lib/js/angular2-polyfills.js:1511:9)
   at lib$es6$promise$$internal$$invokeCallback (http://localhost:8000/lib/js/angular2-polyfills.js:1523:9)
   at Anonymous function (http://localhost:8000/lib/js/angular2-polyfills.js:2071:13)
   at Anonymous function (http://localhost:8000/lib/js/angular2-polyfills.js:243:5)"
   }

**Comment by kneuse** [comment]
The Internet Explorer version is 11 (11.0.9600.18125CO) with rendering engine set to Edge.

**Comment by VeselovAndrey** [comment]
Same for me. Here is the example in Plunker. It will works in IE 11/10 only if you change all "beta.1" to "beta.0" 
http://plnkr.co/edit/wqvzQ4Ma4F1avHqwsIUd?p=preview

**Comment by kneuse** [comment]
I am not sure how you found bug #213, I specifically tried "The use of a keyword for an identifier is invalid" as a search criteria and found nothing. Anyways, I updated the title just in-case the body of these bug reports are not very searchable.

**Comment by kneuse** [comment]
I see why I never found it, because I searched angular/angular rather than angular.

**Comment by alexi2014** [comment]
Same with me. And if my root component contains another component in its view markup then I also get the "SyntaxError: Expected ';'" exception. This happens only in IE11 (ok in Chrome/FF/Edge) and it was not the issue in Beta.0.
The "The use of a keyword for an identifier is invalid" happens in the 'evalExpression' function in its last line starting with 'return new (Function......', line 452 in angular2.dev.js.

**Comment by dtabuenc** [comment]
I think the `vars` argument array is suspect. While debugging in IE 11  I seed keys such as:

`function RouterLink(_router, _location) {\r"`
and
`function RouterOutlet(_elementRef, _loader, _parentRouter, nameAttr) {\r`

I don't know how or why they are there, but these keys are then used to populate the `fnArgNames` array which are used to create the new function arguments hence triggering the "The use of a keyword for an identifier is invalid", which I assume is because we are trying to named a function parameter `function RouterLink(_router,_location){\r`

**Comment by dtabuenc** [comment]
I further traced it down to directive name in the metadata being incorrect. For some reason in IE11 `dirMmeta.type.name` for a router directive is being set to `function RouterLink(_router, _location) {\r"` while in chrome it is simply `RouterLink`

**Comment by dtabuenc** [comment]
While trying to figure out why the compiler was creating incorrect directive metadata, the issue just suddenly fixed itself on my computer and everything works. For those who are still experiencing the issue, can you try fully clearing cache or using incognito mode to see if it still reproduces? Anybody have any insight as to why the type name of a directive would include the first line of the entire function declaration?

**Comment by eelwayadmin** [comment]
Same here for me. I got just a simple view, and that fails on IE11.

**Comment by cyberstorm256** [comment]
I tried your example from plunker, it works in Chrome but not in IE (11.0.9600.18097). Having the same issue "SyntaxError: The use of a keyword for an identifier is invalid" on "angular2.dev.js:452:5".

Clearing IE cache didn't solve the problem.

**Comment by Tragetaschen** [comment]
Dup of #6450?

**Comment by dlebedynskyi** [comment]
+1. beta.0 is fine thought.

**Comment by zybreak** [comment]
+1
Downgrading to beta.0 fixed it for me as well

**Comment by Tragetaschen** [comment]
Ok, mine is the dup :)

Let me just quote #6450 for reference:

In beta.1 (so 7ae23ad is likely the culprit) in IE11, the generated JavaScript from jitting contains syntax errors:

```js
var appProtoView0_function SvgDefsImpl() {}0 = AppProtoView.create(resolvedMetadataCache, 1, [], {});
var renderType56_function SvgDefsImpl() {} = null;
var styles57_function SvgDefsImpl() {} = styles;
```

Instead of a sanitized name, `function SvgDefsImpl() {}` appears in the variable names on virtually every line.

**Comment by sanbornc** [comment]
I found that in lang.ts, the **stringify** function returns the entire first line of the token when the token.name is undefined (IE does not support function.name) so the returned value is *function functionName(arg) {*   Therefore the identifier is named function which is of course a keyword.  

I have verified that the problem is fixed if  I add the following:

    if(typeof token === 'function') {
        return /^function\s+([\w\$]+)\s*\(/.exec(token.toString())[1];
    }

to the stringify function the problem no longer occurs, but it seems like it might be more appropriate to add create a IE specific shim to address the missing name.

**Comment by mlaval** [comment]
We have this shim in the repo as `function.name` is used in some tests:
https://github.com/angular/angular/blob/master/modules/angular2/src/testing/shims_for_IE.js#L1-L14
Does it solve the issue if it is added to the application?

**Comment by justindujardin** [comment]
@mlaval I have been using your IE shims file for my tests on sauce labs, and my builds are still broken on all versions of IE with Beta 1. 

I'm not sure I can create a Plunkr of it, but you can see my Travis/Sauce sadness here: https://travis-ci.org/justindujardin/ng2-material/builds/103002845

Here's the (sparse) stack trace information:

```
LOG: SyntaxError{description: 'Expected ')'', number: -2146827282, stack: 'SyntaxError: Expected ')'
   at evalExpression (Unknown script code:427:5)
   at _createViewFactoryRuntime (Unknown script code:179:13)
   at Anonymous function (Unknown script code:140:21)
   at run (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:120:7)
   at zoneBoundFn (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:93:7)
   at lib$es6$promise$$internal$$tryCatch (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:1493:9)
   at lib$es6$promise$$internal$$invokeCallback (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:1505:9)
   at Anonymous function (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:2053:13)
   at Anonymous function (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:225:5)
   at run (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:120:7)
   at zoneBoundFn (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:93:7)
   at lib$es6$promise$asap$$flush (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:1287:9)
   at Anonymous function (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:475:11)
   at Anonymous function (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:100:7)
   at run (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:120:7)
   at zoneBoundFn (http://localhost:9876/base/node_modules/zone.js/dist/zone-microtask.js?2c40f1490bd97f0e177fa1ec2a3c3000d0f08bd4:93:7)'}
```

**Comment by sanbornc** [comment]
Adding the function name shim in 
https://github.com/angular/angular/blob/master/modules/angular2/src/testing/shims_for_IE.js#L1-L14
Solves the issue for me.

**Comment by BenoitDelrue** [comment]
It seems using beta-0 for angular2.dev.js solve all problems with IE9-10-11.

**Comment by deadZedd** [comment]
As @sanbornc commented, adding the following lines in a script tag or JS file before the loading of any other libraries will fix the problem for IE. (as a temporary fix for beta.1)

```
if (!Object.hasOwnProperty('name')) {
  Object.defineProperty(Function.prototype, 'name', {
    get: function() {
      var matches = this.toString().match(/^\s*function\s*(\S*)\s*\(/);
      var name = matches && matches.length > 1 ? matches[1] : "";
      Object.defineProperty(this, 'name', {value: name});
      return name;
    }
  });
}
```

**Comment by hdorgeval** [comment]
Several days that I was struggling to deploy the QuickStart App (Beta1) to my Windows Phone with Apache Cordova and Visual Studio 2015 : still getting the same error as above. 
Fixed by adding the shims_for_IE.js in the index.html of the Apache Cordova project.
Thanks to @mlaval

**Comment by qdb** [comment]
thanks @sanbornc and @deadZedd for tmp fix, works great

**Comment by glennweb** [comment]
Yip, that @sanbornc and @deadZedd . Fix worked!

**Comment by levyuk** [comment]
Any updates on this issue? Using beta2 it still doesn't work. Nor does having the shim solve the problem or the solution proposed by @deadZedd

**Comment by Queatz** [comment]
I'm getting this, too.  Needs to be fixed.

**Comment by garethflowers** [comment]
This is still not resolved in beta2 - but is fixed by applying the suggested shim (@sanbornc) after.

**Comment by user414** [comment]
Thanks @mlaval the shim solved the issue for me as well using beta2.

**Comment by mlaval** [comment]
There are many issues due to shims being needed in some old browsers: IE11 and below, Android 4.3 and below.
A workaround is to add the ones that are used for unit testing the framework: https://github.com/angular/angular/blob/master/modules/angular2/src/testing/shims_for_IE.js

But they maybe should be somehow published:
- either in `angular2-polyfills.js`, but it means adding 2kb for many all browsers
- or in a new `angular2-shims.js` file, but it will have to be well documented
- or something else

To be discussed

**Comment by justindujardin** [comment]
I looked into this a bit more today and figured out why my build is failing and others are working. 

The IE shim that @mlaval points has a regular expression that matches too liberally.  When I instrument my code with istanbul, the function names get mangled and the Regexp does not match.  

Here's the debug output that points to the regexp failure:
```
LOG: 'fn name -> function MdAnchor(){__cov_b2z45wAFTwiW6Nh3Mukb$w.f['12']++;__cov_b2z45wAFTwiW6Nh3Mukb$w.s['47']++;_super.apply(this,arguments);}'
LOG: 'returns -> MdAnchor(){__cov_b2z45wAFTwiW6Nh3Mukb$w.f['12']++;__cov_b2z45wAFTwiW6Nh3Mukb$w.s['47']++;_super.apply'
```

The full IE name shim looks like this now, and my build is happy:
```javascript
// function.name (all IE)
/*! @source http://stackoverflow.com/questions/6903762/function-name-not-supported-in-ie*/
if (!Object.hasOwnProperty('name')) {
  Object.defineProperty(Function.prototype, 'name', {
    get: function() {
      var matches = this.toString().match(/^\s*function\s*(\S[^\(]*)\s*\(/);
      var name = matches && matches.length > 1 ? matches[1] : "";
      // For better performance only parse once, and then cache the
      // result through a new accessor for repeated access.
      Object.defineProperty(this, 'name', {value: name});
      return name;
    }
  });
}
```

**Comment by alexandrovq** [comment]
Shim for IE made it work for me.
https://github.com/angular/angular/blob/master/modules/angular2/src/testing/shims_for_IE.js

**Comment by theodorejb** [comment]
This error is still occurring in beta 6. Is there an official recommendation for how to work around the issue?

**Comment by CaptainCodeman** [comment]
@theodorejb  Did you try the _modified_ shim that @justindujardin posted above? That fixed the error for me.

**Comment by mikereinhold** [comment]
Modified custom shim also works for me - will the Angular team be providing an office fix for this?

**Comment by whelanp** [comment]
Can confirm https://github.com/angular/angular/blob/master/modules/angular2/src/testing/shims_for_IE.js worked for me with beta6. I was getting the The use of a keyword for an identifier is invalid since beta1

**Comment by theodorejb** [comment]
@CaptainCodeman The https://github.com/angular/angular/blob/master/modules/angular2/src/testing/shims_for_IE.js shim also worked for me with beta 6. Ideally Angular 2 would work with IE 11 out-of-the-box, though, and if not this shim should be included in the npm package and clearly documented.

**Comment by CaptainCodeman** [comment]
@theodorejb Definitely, but sometimes we need a quick fix to unblock us until the proper solution is released and available.

I was having a problem even with the original shim and needed the modified one to fix it.

One of the problems is there are just so many variations on build and deploy - we should maybe start labelling our comments to indicate what we're using.

**Comment by EdmondWang** [comment]
Same issue facing when I use beta.6.
Below shim works for me too.
https://github.com/angular/angular/blob/master/modules/angular2/src/testing/shims_for_IE.js

**Comment by user414** [comment]
FYI for everyone, only the file of @justindujardin works for me in all case, I was still getting some error with @mlaval file that I didn't think was related to that shim but after much time investigating turns out the fix done by @justindujardin fixed it. Maybe @justindujardin can send a pull request so at least the testing shim get fixed in master, in the mean time people can just make sure they use the right file.

**Comment by user414** [comment]
Just to provide info if it helps someone else, my problem was caused by using uglifyjs when setting beautify to false I would get

```
EXCEPTION: SyntaxError: Expected ')'
```
in IE 11, but would work fine when beautify was set to true. I wasn't using istanbul so never though to change the shim to @justindujardin versions, until I looked deeper and realized the regexp caused the problem as well for me.

**Comment by jared83** [comment]
I had the same issue. The shim didn't work right away for me, as in my project I was both, minifying and mangling js. Once I disabled them in my gulpfile, the shim worked.
The details on the issue are here #6695 
Anyone knows is that issue has been addressed in beta.3?

**Comment by user414** [comment]
@jared83 Reread my post and it wasn't clear, with the shim from @justindujardin I can use uglifyjs to compress and minify, the mangling is a separate issue has to do with the angular generated min file not working like angular2.min.js

https://github.com/angular/angular/issues/6380

**Comment by KevinJannis** [comment]
Fixed it with adding https://github.com/angular/angular/blob/master/modules/angular2/src/testing/shims_for_IE.js#L1-L14 to the code.

**Comment by user414** [comment]
@KevinJannis As mentionned previously that's not the best fix and the one that should be used IMHO, the one from @justindujardin with the regexp fixed is the one.

**Comment by KevinJannis** [comment]
@user414 Thanks, will update to the one with the fix for regexp.

**Comment by user414** [comment]
FYI linking the pull request here, it should be merged in soon hopefully

https://github.com/angular/angular/pull/6944

**Comment by ronakvbhatt07** [comment]
working with **beta.7** and used **shims_for_IE.js** file with it, that resolved basic issue that now we can access pages in IE. But now we are facing IE designing issue and the other one is Images gets disappeared.

**Comment by jkristia** [comment]
I'm not sure if this is fixed or not (hard to follow the thread), but I see the same issue on IE 11 (11.162.10586.0 Win 10). 2.0.0-beta.7

**Comment by pranavgawri** [comment]
As suggested, fixed it by adding https://github.com/angular/angular/blob/master/modules/angular2/src/testing/shims_for_IE.js#L1-L14 to the code.

**Comment by SethDavenport** [comment]
Any particular reason these shims aren't part of angular-polyfills?  I'm a bit reluctant to pull code from the bowels of the lib since it feels 'unofficial'. Also the current `shims_for_IE` doesn't work very well with webpack.  I'll use a `<script>` tag for now but guidance on the 'correct' approach would be appreciated.

**Comment by ericmartinezr** [comment]
@SethDavenport that's being discussed over here https://github.com/angular/angular/pull/7390

**Comment by user414** [comment]
@SethDavenport, people over at angular webpack starter project have made a 

https://github.com/gdi2290/ie-shim
https://github.com/AngularClass/angular2-webpack-starter

project that can be integrated in webpack. If you look at their config file they use it in there. However, it doesn't have the @justindujardin version but the original one with causes some problem with the regexp. I'm assuming @gdi2290 might be waiting for the pull reference above before updating that file... I personally just patch the code myself until it all gets squared off.

**Comment by SethDavenport** [comment]
Thanks, I'll check it out.  In my case it turned out to tsconfig settings that were messing webpack up.

**Comment by SethDavenport** [comment]
@user414 FWIW I ended up doing this: https://github.com/rangle/angular2-redux-starter/pull/81/files

* Got similar polyfills for `classlist`, `requestAnimationFrame`, and `console` from `npm`: probably a bit more code but with better packaging and more clarity on licensing.
* Added an additional polyfill for `history.pushState` which is needed for IE9.
* Adopted @justindujardin's variant of `function.name` for now.

Thanks for the tips.

**Comment by alexeagle** [comment]
Looks to me like there are a few issues lumped together here, can someone clarify?
- IE doesn't implement `function.name`
- angular 2 stringifies things into a form that contains keywords like 'proto'
- angular 2 stringifies functions into `function (arg) {`

In #7825 I fixed the latter one, only in the case of the template compiler doing codegen (because we found that closure compiler drops class names when lowering es6 to es5). If you're using TypeScript or Babel to produce ES5, you don't have this problem.

Also my fix produces an opaque identifier like `anonymous_function_0_` which means debugging is harder. We should avoid that as a general solution, and try to get the name of the function (or class) to propagate all the way to the string form.

**Comment by CaptainCodeman** [comment]
My IE experience:

I use the modified IE shim that @justindujardin posted [above](https://github.com/angular/angular/issues/6501#issuecomment-181145904)

Other than that, everything else come from a polyfill service (to also support Safari with Intl) and core.js (which is only required by IE).

The only pieces of core.js actually needed by IE appear to be:

core-js/es6/array
core-js/es6/map
core-js/es6/string
core-js/es6/symbol

But this was pre-zone 0.6.8 / beta.13 fix where I pull in zone directly instead of using angular2-polyfills so core.js is also supplying reflect-metadata as well so I just pull in core.js from a CDN for the time being.

Everything is built with webpack and uglified but without mangle. Source here in case it helps anyone: https://github.com/CaptainCodeman/angular2-webpack-polyfill

**Comment by gdi2290** [comment]
here's the mangle config to get everything working with minify 
```coffee
mangle: {
  screw_ie8 : true,
  keep_fnames: true
}
```

**Comment by CaptainCodeman** [comment]
> here's the mangle config to get everything working with minify

Thanks you so much - that works great !!

**Comment by mhevery** [comment]
Fixed by #6944

**Comment by dlebedynskyi** [comment]
`kneuse commented on Jan 15` 
@mhevery closes on May 20.
label for issue still `effort1: easy (hour)`.
Thanks for fix!
